### ISSUE TITLE
sharing from Firefox does not work

### ISSUE THREAD

**Issue by rgh36167** [theScrabi/NewPipe#84]
When sharing youtube links from Firefox share menu NewPipe isn't offered as an option, apparently the intent pattern does not match.

Examined by IntentIntercept the intent looks like this:

```
intent:#Intent;action=android.intent.action.SEND;type=text/plain;launchFlags=0x10080001;B.org.mozilla.firefox.intent.extra.DEVIC

intent:#Intent;action=android.intent.action.SEND;type=text/plain;launchFlags=0x10080001;B.org.mozilla.firefox.intent.extra.DEVICES_ONLY=true;S.android.intent.extra.SUBJECT=ZDF%20-%20Islam%20-
%20Effects%20on%20Germany%20-%20YouTube;S.android.intent.extra.TEXT=https%3A%2F%2Fm.youtube.com%2F%23%2Fwatch%3Fv%3DKVWAIKoatWM;S.android.intent.extra.TITLE=ZDF%20-%20Islam%20-%20Effects%20on
%20Germany%20-%20YouTube;end

ACTION: android.intent.action.SEND
DATA: null
TYPE: text/plain

FLAGS:
FLAG_ACTIVITY_CLEAR_WHEN_TASK_RESET
FLAG_GRANT_READ_URI_PERMISSION
FLAG_RECEIVER_FOREGROUND

EXTRAS:
EXTRA 1:
Class: java.lang.Boolean
Key: org.mozilla.firefox.intent.extra.DEVICES_ONLY
Value: true
EXTRA 2:
Class: java.lang.String
Key: android.intent.extra.SUBJECT
Value: ZDF - Islam - Effects on Germany - YouTube
EXTRA 3:
Class: java.lang.String
Key: android.intent.extra.TEXT
Value: https://m.youtube.com/#/watch?v=KVWAIKoatWM
EXTRA 4:
Class: java.lang.String
Key: android.intent.extra.TITLE
Value: ZDF - Islam - Effects on Germany - YouTube


14 ACTIVITIES MATCH THIS INTENT:
Messaging (com.android.mms - com.android.mms.ui.ComposeMessageActivity)
Bluetooth (com.android.bluetooth - com.android.bluetooth.opp.BluetoothOppLauncherActivity)
Hangouts (com.google.android.talk - com.google.android.apps.hangouts.phone.ShareIntentActivity)
Translate (com.google.android.apps.translate - com.google.android.apps.translate.TranslateActivity)
Tasks (org.tasks - com.todoroo.astrid.activity.ShareLinkActivity)
Add to Firefox (org.mozilla.firefox - org.mozilla.gecko.overlays.ui.ShareDialog)
BlitzMail (de.grobox.blitzmail - de.grobox.blitzmail.SendActivity)
Barcode Scanner (com.google.zxing.client.android - com.google.zxing.client.android.encode.EncodeActivity)
920 Text Editor (com.jecelyin.editor - com.jecelyin.editor.JecEditor)
Encrypt (org.thialfihar.android.apg - org.thialfihar.android.apg.ui.EncryptActivity)
Decrypt (org.thialfihar.android.apg - org.thialfihar.android.apg.ui.DecryptActivity)
Skype (com.skype.raider - com.skype.android.app.main.SplashActivity)
K-9 Mail (com.fsck.k9 - com.fsck.k9.activity.MessageCompose)
Term here (jackpal.androidterm - jackpal.androidterm.TermHere)
```

**Comment by medavox** [comment]
I've just looked at the manifest file; apparently NewPipe doesn't currently declare itself as handling the android.intent.action.SEND intent.

I can add this fairly easily, I think.

Thanks for the useful info with IntentIntercept; it made diagnosis a lot easier for me :)

**Comment by medavox** [comment]
OK, I've looked at the android Intent Filter docs, and this is a pretty badly formed Intent.

The Data field shouldn't be null; I can't perform any useful Intent filtering without it. If I register NewPipe to handle `android.intent.action.SEND` for mime type `text/plain`, it'll appear inappropriately a lot of the time in share menu, for data it can't really handle.

Filtering by Extra (which is where the URL is stored in this Intent) doesn't seem to be possible in the Manifest. AFAICT, there'd have to be a bunch of validation routines that checked the Intent after NewPipe formally accepted it, and launched. If the Intent was not a URL or was malformed (eg using another EXTRA field name for the URL), NewPipe would have been launched for nothing.

In short, I don't think NewPipe should catch this intent. Although obviously this is completely open for discussion.

**Comment by rgh36167** [comment]
Dentex Youtube downloader did it so it is possible. I can't find the latest sources but look at this: https://github.com/surahul/ytdownloader-1/blob/master/dentex.youtube.downloader/AndroidManifest.xml

Apparently it does simply 
```
<intent-filter>
                <action android:name="android.intent.action.SEND" />

                <category android:name="android.intent.category.DEFAULT" />

                <data android:mimeType="text/plain" />
            </intent-filter>
```
and  nobody complained that the extra entry in the share menu won't make sense in most cases. People are quite used that 90% of the options offered in the share menu don't make little sense. Maybe there is some better way to do it but I don't know of any.

Of course I am wondering why Firefox uses SEND to share links, there would seem to be much better methods? Maybe it is possible to configure Firefox for Android somehow?

**Comment by andreicristianpetcu** [comment]
Please add this feature! This is really usefull. Thank you!

**Comment by medavox** [comment]
OK, I will have another look at this after I've implemented [timestamps](https://github.com/theScrabi/NewPipe/issues/59).

In the meantime, I welcome anyone else to implement this :)

**Comment by theScrabi** [comment]
It's no big deal I'll do that.

**Comment by medavox** [comment]
@theScrabi awesome. It turns out someone also requested this on NewPipe's F-droid [forum thread](https://f-droid.org/forums/topic/newpipe-2/#post-17853)

**Comment by theScrabi** [comment]
https://github.com/theScrabi/NewPipe/issues/101

**Comment by andreicristianpetcu** [comment]
It works for me now. v0.6.2

**Comment by theScrabi** [comment]
Did someone implement this? I didn't O.o.

**Comment by andreicristianpetcu** [comment]
Apparently Firefox shows in the location bar a little android icon. If I click that icon it opens up the native app and it works for all native apps like Reddit, Github or Tinfoil for Facebook. It's not in the share menu, but I don't want that. I just want to be able to jump from Firefox to NewPipe.

**Comment by theScrabi** [comment]
You may be right, but even I think it would be nice to open a video in NP through share from another app.

**Comment by eighthave** [comment]
I don't think *Share* aka `ACTION_SEND` is the right thing to do this, it is meant for sending data like as in a message.  It is not meant for viewing something in a different app aka `ACTION_VIEW`, which is what is being described here.  Sure its possible to implement this in NewPipe, but that would clutter up the Android user experience.

In the case of Firefox, if you long press on a URL, like a movie in youtube.com, then you can select "Open With an App" and then choose NewPipe.

**Comment by Thomas131** [comment]
the share-function would be really nice! Sadly, I don't have much Android-programming-knowledge, so I can't help.

**Comment by eighthave** [comment]
I got stuck into an odd place with trying to get a video in a website in
firefox to play in NewPipe, and it seems that Share might be the only
way.  Firefox's odd VIEW/SEND Intent handling is frustrating but
something that probably needs to be worked around...

**Comment by theScrabi** [comment]
Sounds like a go. I'll keep it in mind.

**Comment by Thomas131** [comment]
Thanks so much for making the share-function! This makes NewPipe my default Youtube-App!

**Comment by theScrabi** [comment]
So I think I've came up with a solution. I offer share function as an external app. Therefore you can decide yourself, if you want to have this function or not:
https://github.com/theScrabi/ShareWithNewPipe
